R/convenience-functions.R

```diff
@@ -0,0 +1,194 @@
+#'
+#' # specifying an offset manually for the log transformation removes the warning
+#' transform_forecasts(example_quantile, truncate = TRUE, offset = 1)
+#'
+#' # adding multiple transformations
+#' library(magrittr) # pipe operator
+#' example_quantile %>%
+#'   transform_forecasts(offset = 1, truncate = TRUE) %>%
+#'   # manually truncate all negative values before applying sqrt
+#'   transform_forecasts(fun = function(x) pmax(0, x), append = FALSE, label = "natural") %>%
+#'   transform_forecasts(fun = sqrt, label = "sqrt") %>%
+#'   score() %>%
+#'   summarise_scores(by = c("model", "scale"))
+
+transform_forecasts <- function(data,
+                                fun = log_shift,
+                                append = TRUE,
+                                label = "log",
+                                ...) {
+
+  original_data <- as.data.table(data)
+  scale_col_present <- ("scale" %in% colnames(original_data))
+
+  # Error handling
+  if (scale_col_present) {
+    if (!("natural" %in% original_data$scale)) {
+      stop("If a column 'scale' is present, entries with scale =='natural' are required for the transformation")
+    }
+    if (append && (label %in% original_data$scale)) {
+      w <- paste0(
+        "Appending new transformations with label '", label,
+        "', even though that entry is already present in column 'scale'."
+      )
+      warning(w)
+    }
+  }
+
+  if (append) {
+    if (scale_col_present) {
+      transformed_data <- data.table::copy(original_data)[scale == "natural"]
+    } else {
+      transformed_data <- data.table::copy(original_data)
+      original_data[, scale := "natural"]
+    }
+    transformed_data[, prediction := fun(prediction, ...)]
+    transformed_data[, true_value := fun(true_value, ...)]
+    transformed_data[, scale := label]
+    out <- rbind(original_data, transformed_data)
+    return(out[])
+  }
+
+  # check if a column called "scale" is already present and if so, only
+  # restrict to transformations of the original data
+  if (scale_col_present) {
+    original_data[scale == "natural", prediction := fun(prediction, ...)]
+    original_data[scale == "natural", true_value := fun(true_value, ...)]
+    original_data[scale == "natural", scale := label]
+  } else {
+    original_data[, prediction := fun(prediction, ...)]
+    original_data[, true_value := fun(true_value, ...)]
+  }
+  return(original_data[])
+}
+
+
+
+
+
+
+#' @title Log transformation with an additive shift
+#'
+#' @description Function that shifts a value by some offset and then applies the
+#' natural logarithm to it.
+#'
+#' @details The output is computed as log(x + offset) (or
+#' log(pmax(0, x) + offset)) if `truncate = TRUE`.
+#'
+#' @param x vector of input values to be transformed
+#' @param offset number to add to the input value before taking the natural
+#' logarithm
+#' @param truncate whether to truncate negative values to 0. Default is FALSE.
```

Comment:
I don't know what this means before looking at the code. Does the truncation happen before or after the transform, and why would one want to truncate?